### PR TITLE
[gql][consistent-reads][2/n] Implement consistent reads for Balance

### PR DESCRIPTION
**Pull request by wlmyng**
## Description 

Implements consistent reads for the Balance type.

1. RawQuery struct builds a BoxedSqlQuery to support consistent reads. The query demands more than what diesel's dsl can provide today, so unfortunately we will have it in raw sql for the time being. 
2. Split the pagination methods of the Target trait into a separate, Paginated trait. Implement target to build a cursor for a database entity.
3. Break the validation step out from paginate_query, so that the new paginate_raw_query method, which accepts a thunk that builds a RawQuery, can also have its results validated.

## Test Plan 

consistency/balances.move

---
If your changes are not user-facing and do not break anything, you can skip the following section. Otherwise, please briefly describe what has changed under the Release Notes section.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-graphql-rpc/src/types/balance.rs [hunk 1]
```diff
@@ -80,27 +79,35 @@ impl Balance {
...
+                        return Ok(None);
+                    }
+                    rhs = checkpoint_viewed_at;
+                }
+
+                let result = page.paginate_raw_query::<StoredBalance>(conn, move || {
+                    Balance::base_query(address, None, lhs as i64, rhs as i64)
+                })?;
+
+                Ok::<_, diesel::result::Error>(Some(result))
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/balance.rs:99]: I'm surprised that you need to annotate the error here -- what happens when you leave that off?

File: crates/sui-graphql-rpc/src/types/balance.rs [hunk 2]
```diff
@@ -110,37 +117,75 @@ impl Balance {
 
         Ok(conn)
     }
-}
 
-impl Target<Cursor> for StoredBalance {
-    type Source = objects::table;
+    fn base_query(address: SuiAddress, coin_type: Option<TypeTag>, lhs: i64, rhs: i64) -> RawQuery {
```
[the remark this answers is not in the document]
Comment (amnn) [reply crates/sui-graphql-rpc/src/types/balance.rs:121]: + copious docs please

File: crates/sui-graphql-rpc/src/types/balance.rs [hunk 3]
```diff
@@ -110,37 +117,75 @@ impl Balance {
...
+            format!(r#"checkpoint_sequence_number BETWEEN {} AND {}"#, lhs, rhs)
+        );
+
+        let final_ = consistent_object_read(snapshot_objs, history_objs, lhs, rhs);
+        final_.group_by("candidates.coin_type")
+    }
 
-    fn filter_ge<ST, GB>(cursor: &Cursor, query: Query<ST, GB>) -> Query<ST, GB> {
-        query.filter(objects::dsl::coin_type.ge((**cursor).clone()))
+    fn filter(mut query: RawQuery, address: SuiAddress, coin_type: Option<TypeTag>) -> RawQuery {
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/balance.rs:136]: Ditto (top-level function + docs).

File: crates/sui-graphql-rpc/src/types/balance.rs [hunk 4]
```diff
@@ -110,37 +117,75 @@ impl Balance {
...
-        if asc {
-            query.order_by(dsl::coin_type.asc())
-        } else {
-            query.order_by(dsl::coin_type.desc())
+    fn filter_le(cursor: &Cursor, query: RawQuery) -> RawQuery {
+        filter!(query, "candidates.coin_type <= {}", (**cursor).clone())
+    }
+
+    fn order(asc: bool, query: RawQuery) -> RawQuery {
+        match asc {
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/balance.rs:167]: This is a really strange way to write `if`.

File: crates/sui-graphql-rpc/src/types/balance.rs [hunk 5]
```diff
@@ -110,37 +117,75 @@ impl Balance {
...
-    fn filter_ge<ST, GB>(cursor: &Cursor, query: Query<ST, GB>) -> Query<ST, GB> {
-        query.filter(objects::dsl::coin_type.ge((**cursor).clone()))
+    fn filter(mut query: RawQuery, address: SuiAddress, coin_type: Option<TypeTag>) -> RawQuery {
+        query = filter!(
+            query,
+            format!(
+                "owner_id = '\\x{}'::bytea AND coin_type IS NOT NULL",
+                hex::encode(address.into_vec())
+            )
+        );
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/balance.rs:143]: nit: separate `filter!` calls?

```suggestion
        query = filter!(query, "coin_type IS NOT NULL");
        query = filter!(query, format!("owner_id = '\\x{}'::bytea", hex::encode(address.into_vec())));
```

File: crates/sui-graphql-rpc/src/types/balance.rs [hunk 6]
```diff
@@ -110,37 +117,75 @@ impl Balance {
...
+
+        query
     }
+}
 
-    fn filter_le<ST, GB>(cursor: &Cursor, query: Query<ST, GB>) -> Query<ST, GB> {
-        query.filter(objects::dsl::coin_type.le((**cursor).clone()))
+impl RawPaginated<Cursor> for StoredBalance {
+    fn filter_ge(cursor: &Cursor, query: RawQuery) -> RawQuery {
+        filter!(query, "candidates.coin_type >= {}", (**cursor).clone())
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/balance.rs:159]: Probably worth adding a comment here explaining where "candidates" comes from.

Comment (amnn) [reply crates/sui-graphql-rpc/src/types/balance.rs:159]: Although actually, do you need that at all? (e.g. could you get away with not specifying `candidates.`?

File: crates/sui-graphql-rpc/src/types/balance.rs [hunk 7]
```diff
@@ -160,3 +205,54 @@ impl TryFrom<StoredBalance> for Balance {
...
     }
 }
+
+pub(crate) fn build_candidates(snapshot_objs: RawQuery, history_objs: RawQuery) -> RawQuery {
+    let candidates = query!(
+        r#"SELECT DISTINCT ON (object_id) * FROM (
+        ({})
+        UNION
+        ({})
+    ) o"#,
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/balance.rs:215]: nit: this could now be one line?

```suggestion
        r#"SELECT DISTINCT ON (object_id) * FROM (({}) UNION ({})) o"#,
```

File: crates/sui-graphql-rpc/src/types/balance.rs [hunk 8]
```diff
@@ -160,3 +205,54 @@ impl TryFrom<StoredBalance> for Balance {
...
+pub(crate) fn build_join(candidates: RawQuery, newer: RawQuery) -> RawQuery {
+    let final_ = query!(
+        r#"
+    SELECT CAST(SUM(candidates.coin_balance) AS TEXT) as balance, COUNT(*) as count, candidates.coin_type as coin_type
+        FROM ({}) candidates
+        LEFT JOIN ({}) newer
+        ON (
+            candidates.object_id = newer.object_id
+            AND candidates.object_version < newer.object_version
+        )"#,
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/balance.rs:242]: nit: whitespace.

```suggestion
        r#"SELECT 
            CAST(SUM(candidates.coin_balance) AS TEXT) as balance, 
            COUNT(*) as count, 
            candidates.coin_type as coin_type
        FROM ({}) candidates
        LEFT JOIN ({}) newer
        ON (
            candidates.object_id = newer.object_id
            AND candidates.object_version < newer.object_version
        )"#,
```

Also on my earlier question of whether you could avoid mentioing `candidates.` in the cursor query, I think the answer is yes, if you also join on `candidates.coin_type = newer.coin_type` which should be true because coin types are not going to change between different versions of the same object (we hope).

Comment (amnn) [reply crates/sui-graphql-rpc/src/types/balance.rs:242]: I also have a little bit of a hard time figuring out why the query building is split up into multiple functions. I can see the benefit of doing that for the innermost `filter` application, but otherwise it's making things quite hard to follow, because I'm jumping around.

Maybe I'm missing a place where these things are being re-used, in which case feel free to ignore me, but if it's purely to split things up then I would recommend opting to have them in a single function, but use comments to split that up into recognisable sections.

File: crates/sui-graphql-rpc/src/types/cursor.rs [hunk 9]
```diff
@@ -74,7 +75,17 @@ pub(crate) trait Target<C: CursorType> {
     /// (returning the new query). The `asc` parameter controls whether the ordering is ASCending
     /// (`true`) or descending (`false`).
     fn order<ST, GB>(asc: bool, query: Query<ST, Self::Source, GB>) -> Query<ST, Self::Source, GB>;
+}
+
+pub(crate) trait RawPaginated<C: CursorType>: Target<C> {
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/cursor.rs:80]: Docs

File: crates/sui-graphql-rpc/src/types/cursor.rs [hunk 10]
```diff
@@ -74,7 +75,17 @@ pub(crate) trait Target<C: CursorType> {
...
+    fn filter_ge(cursor: &C, query: RawQuery) -> RawQuery;
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/cursor.rs:81]: on

File: crates/sui-graphql-rpc/src/types/cursor.rs [hunk 11]
```diff
@@ -74,7 +75,17 @@ pub(crate) trait Target<C: CursorType> {
...
 
+    fn filter_le(cursor: &C, query: RawQuery) -> RawQuery;
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/cursor.rs:83]: everything

File: crates/sui-graphql-rpc/src/types/cursor.rs [hunk 12]
```diff
@@ -74,7 +75,17 @@ pub(crate) trait Target<C: CursorType> {
...
+
+    fn order(asc: bool, query: RawQuery) -> RawQuery;
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/cursor.rs:85]: please

File: crates/sui-graphql-rpc/src/types/cursor.rs [hunk 13]
```diff
@@ -228,67 +239,120 @@ impl<C: CursorType + Eq + Clone + Send + Sync + 'static> Page<C> {
...
+        ))
+    }
 
-            // Similar to above case, but for back of results.
-            (_, _, Some(l), Some(b), End::Back) if l.cursor() != *b => {
-                return Ok((false, false, vec![].into_iter()));
+    pub(crate) fn paginate_raw_query<T>(
+        &self,
+        conn: &mut Conn<'_>,
+        query_fn: impl Fn() -> RawQuery,
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/cursor.rs:252]: I don't think you need `paginate_raw_query` to accept a query thunk, because `RawQuery` can implement `clone` just fine.

File: crates/sui-graphql-rpc/src/types/cursor.rs [hunk 14]
```diff
@@ -228,67 +239,120 @@ impl<C: CursorType + Eq + Clone + Send + Sync + 'static> Page<C> {
...
         };
 
+        Ok(self.paginate_results(
+            results.first().map(|f| f.cursor()),
+            results.last().map(|l| l.cursor()),
+            results,
+        ))
+    }
+
+    fn paginate_results<T>(
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/cursor.rs:291]: Docs for this please!

File: crates/sui-graphql-rpc/src/types/cursor.rs [hunk 15]
```diff
@@ -228,67 +239,120 @@ impl<C: CursorType + Eq + Clone + Send + Sync + 'static> Page<C> {
...
+            results.first().map(|f| f.cursor()),
+            results.last().map(|l| l.cursor()),
+            results,
+        ))
+    }
 
-            // Similar to above case, but for back of results.
-            (_, _, Some(l), Some(b), End::Back) if l.cursor() != *b => {
-                return Ok((false, false, vec![].into_iter()));
+    pub(crate) fn paginate_raw_query<T>(
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/cursor.rs:249]: Docs

**Review by amnn — APPROVED** [review]
The implementation here looks good, but this needs infinity% more documentation, because right now there is essentially none, and this is one of the most complicated parts of the system at the moment!

File: crates/sui-graphql-rpc/src/data.rs [hunk 16]
```diff
@@ -94,3 +107,170 @@ pub(crate) trait DbConnection {
...
+            $select,
+            $({
+                let (sub_sql, sub_binds) = $subquery.finish();
+                binds.extend(sub_binds);
+                sub_sql
+            }),*
+        );
+
+        RawQuery::new(select, binds)
+    }};
```
[the remark this answers is not in the document]
Comment (wlmyng) [reply crates/sui-graphql-rpc/src/data.rs:275]: sounds good to me, thanks!

File: crates/sui-graphql-rpc/src/types/balance.rs (hunk 4 above)
Comment (wlmyng) [reply crates/sui-graphql-rpc/src/types/balance.rs:167]: lol i must be losing my mind

File: crates/sui-graphql-rpc/src/types/balance.rs (hunk 8 above)
Comment (wlmyng) [reply crates/sui-graphql-rpc/src/types/balance.rs:242]: That's fair, I'll put it back into one function.

Comment (wlmyng) [reply crates/sui-graphql-rpc/src/types/balance.rs:242]: And unfortunately, even with the additional join, postgres will still complain that just `coin_type` is ambiguous. Just tried it manually :/

File: crates/sui-graphql-rpc/src/types/balance.rs (hunk 6 above)
Comment (wlmyng) [reply crates/sui-graphql-rpc/src/types/balance.rs:159]: Actually yeah, I can get away with not specifying `candidates`. To do so, we have to make sure `coin_type` is not a column on `newer`.

Comment (wlmyng) [reply crates/sui-graphql-rpc/src/types/balance.rs:159]:
```
                candidates.object_id = newer.object_id
                AND candidates.object_version < newer.object_version
```

Just this bit where we need both aliases. It'll only work for the balance query since the columns selected from `candidates` and `newer` do not overlap.

File: crates/sui-graphql-rpc/src/data.rs [hunk 17]
```diff
@@ -94,3 +107,170 @@ pub(crate) trait DbConnection {
...
+
+        if let Some(first) = group_by.next() {
+            select.push_str(" GROUP BY ");
+            select.push_str(first);
+        }
+
+        for group in group_by {
+            select.push_str(", ");
+            select.push_str(group);
+        }
```
[the remark this answers is not in the document]
Comment (wlmyng) [reply crates/sui-graphql-rpc/src/data.rs:179]: Less verbose for sure. I'll adopt it

File: crates/sui-graphql-rpc/src/types/balance.rs (hunk 1 above)
Comment (amnn) [reply crates/sui-graphql-rpc/src/types/balance.rs:101]: Yeah, I have no idea what's going on here either -- there's even prior calls to `Ok` in the same function?  Only suggestion I have is to try a slightly different pattern:

```rust
page.paginate_raw_query::<StoredBalance>(
    conn,
    balance_query(address, None, lhs as i64, rhs as i64),
)
.map(Some);
```

File: crates/sui-graphql-rpc/src/raw_query.rs [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+///    `sql_query.bind`.
+///
+/// 2. **Query String Merging**: Can be used to represent and merge query strings and their
+///    associated bindings. Placeholder strings and bindings are applied in sequential order.
+///
+/// Note: `RawQuery` only supports binding string values, as interpolating raw strings directly
+/// increases exposure to SQL injection attacks.
+#[derive(Clone)]
+pub(crate) struct RawQuery {
+    /// The `SELECT` clause of the query.
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/raw_query.rs:27]: nit: (and the `FROM` clause).

File: crates/sui-graphql-rpc/src/raw_query.rs [hunk 19]
```diff
@@ -0,0 +1,199 @@
...
+        self.limit = Some(limit);
+        self
+    }
+
+    /// Adds the `String` value to the list of binds for this query.
+    pub(crate) fn bind_value(&mut self, condition: String) {
+        self.binds.push(condition);
+    }
+
+    /// Constructs the query string and returns it along with the list of binds for this query.
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/raw_query.rs:98]: Worth calling out that this isn't intended for people to call under normal operation, but is exposed for the `query!` macro to use?

File: crates/sui-graphql-rpc/src/raw_query.rs [hunk 20]
```diff
@@ -0,0 +1,199 @@
...
+
+        for bind in binds {
+            diesel_query = diesel_query.bind::<diesel::sql_types::Text, _>(bind);
+        }
+
+        diesel_query
+    }
+}
+
+#[macro_export]
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/raw_query.rs:158]: Docs for these as well please, especially to explain the interpolation logic!

File: crates/sui-graphql-rpc/src/types/balance.rs (hunk 1 above)
Comment (wlmyng) [reply crates/sui-graphql-rpc/src/types/balance.rs:101]: Ah yeah that did the trick. It's because the first one returns Result<None>, while the second one is just T, so dropping the `?` and adding the `.map(Some)` conforms to `Result<Option<T>>`
